core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java

```diff
@@ -97,13 +101,23 @@ private static <T> T visitUnion(Type type, Schema union, AvroSchemaWithTypeVisit
         options.add(visit(type, branch, visitor));
       }
     } else { // complex union case
-      int index = 1;
-      for (Schema branch : types) {
-        if (branch.getType() == Schema.Type.NULL) {
-          options.add(visit((Type) null, branch, visitor));
+      Map<Integer, Types.NestedField> icebergFieldMap = new HashMap<>();
```

Comment:
I think you could avoid the map and index bookkeeping if you iterate on both types in sort-merge style in just one loop. I will let you decide if this looks simpler.

Reply:
Good point. I refactor this part of code without the map for bookkeeping and create a new function named `visitComplexUnion` to contain these code.